pep-0677.rst

```diff
@@ -0,0 +1,854 @@
+<https://basarat.gitbook.io/typescript/type-system/callable#arrow-syntax>`_,
+`Kotlin <https://kotlinlang.org/docs/lambdas.html>`_, and `Scala
+<https://docs.scala-lang.org/tour/higher-order-functions.html>`_.
+
+Our goals are that:
+
+- Callable types using this syntax will be easier to learn and use,
+  particularly for developers with experience in other languages.
+- Library authors will be more likely to use expressive types for
+  callables that enable type checkers to better understand code and
+  find bugs, as in the ``decorator`` example above.
+
+Consider this simplified real-world example from a web server, written
+using the existing ``typing.Callable``::
+
+    from typing import Awaitable, Callable
+    from app_logic import Response, UserSetting
+
+
+    async def customize_response_for_settings(
+        response: Response,
+        customizer: Callable[[Response, list[UserSetting]], Awaitable[Response]]
+    ) -> Response:
+       ...
+
+With our proposal, this code can be abbreviated to::
+
+    from app_logic import Response, UserSetting
+
+    def make_endpoint(
+        response: Response,
+        customizer: async (Response, list[UserSetting]) -> Response,
+    ) -> Response:
+        ...
+
+This is shorter and requires fewer imports. It also has far less
+nesting of square brackets - only one level, as opposed to three in
+the original code.
+
+Rationale
+=========
+
+The ``Callable`` type is widely used. For example, `as of October 2021
+it was
+<https://github.com/pradeep90/annotation_collector#overall-stats-in-typeshed>`_
+the fifth most common complex type in typeshed, after ``Optional``,
+``Tuple``, ``Union``, and ``List``.
+
+Most of the other commonly used types have had their syntax improved
+via either PEP 604 or PEP 585. ``Callable`` is used heavily enough to
+similarly justify a more usable syntax.
+
+In this proposal, we chose to support all the existing semantics of
+``typing.Callable``, without adding support for new features. We took
+this decision after examining how frequently each feature might be
+used in existing typed and untyped open-source code. We determined
+that the vast majority of use cases are covered.
+
+We considered adding support for named, optional, and variadic
+arguments. However, we decided against including these features, as
+our analysis showed they are infrequently used. When they are really
+needed, it is possible to type these using `Callback Protocols
+<https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols>`_.
+
+See the Rejected Alternatives section for more detailed discussion
+about omitted features.
+
+Specification
+=============
+
+Typing Behavior
+---------------
+
+Type checkers should treat the new syntax with exactly the same
+semantics as ``typing.Callable``.
+
+As such, a type checker should treat the following pairs exactly the
+same::
+
+    from typing import Awaitable, Callable, Concatenate, ParamSpec, TypeVarTuple
+
+    P = ParamSpec("P")
```

Comment:
Feel like it is worth adding one line before each pair explaining what it is that, say, `() -> bool` means.

Five years from now, if all goes to plan, nobody will even know what `typing.Callable[[], bool]` meant 🤣  

So, it's our job to explain what it means. For complex stuff like ParamSpec, we can at least link to the PEP for more details.